pkg/splunk/enterprise/indexercluster.go

```diff
@@ -118,23 +122,168 @@ func ApplyIndexerCluster(client splcommon.ControllerClient, cr *enterprisev1.Ind
 
 	// no need to requeue if everything is ready
 	if cr.Status.Phase == splcommon.PhaseReady {
+		if len(cr.Status.IndexerSecretChanged) > 0 {
+			// Disable maintenance mode
+			err = SetCmMaintenanceMode(client, cr, false, false)
+			if err != nil {
+				return result, err
+			}
+		}
+
+		// Reset idxc secret changed and namespace secret revision
+		cr.Status.IndexerSecretChanged = []bool{}
+		cr.Status.NamespaceSecretResourceVersion = namespaceScopedSecret.ObjectMeta.ResourceVersion
+
 		result.Requeue = false
 	}
 	return result, nil
 }
 
 // indexerClusterPodManager is used to manage the pods within an indexer cluster
 type indexerClusterPodManager struct {
+	c               splcommon.ControllerClient
 	log             logr.Logger
 	cr              *enterprisev1.IndexerCluster
 	secrets         *corev1.Secret
 	newSplunkClient func(managementURI, username, password string) *splclient.SplunkClient
 }
 
+// SetCmMaintenanceMode enables/disables cluster maintenance mode
+func SetCmMaintenanceMode(c splcommon.ControllerClient, cr *enterprisev1.IndexerCluster, enable bool, mock bool) error {
+	// Retreive admin password from Pod
+	masterIdxcName := cr.GetName()
+	// For multisite / multipart cluster, use the cluster-master API of the referenced IndexerCluster
+	if len(cr.Spec.ClusterMasterRef.Name) > 0 {
+		masterIdxcName = cr.Spec.ClusterMasterRef.Name
+	}
+	cmPodName := fmt.Sprintf("splunk-%s-cluster-master-0", masterIdxcName)
+	adminPwd, err := splutil.GetSpecificSecretTokenFromPod(c, cmPodName, cr.GetNamespace(), "password")
+	if err != nil {
+		return err
+	}
+
+	var command string
+	if enable {
+		command = fmt.Sprintf("/opt/splunk/bin/splunk enable maintenance-mode --answer-yes -auth admin:%s", adminPwd)
+	} else {
+		command = fmt.Sprintf("/opt/splunk/bin/splunk disable maintenance-mode --answer-yes -auth admin:%s", adminPwd)
+	}
+	_, _, err = splutil.PodExecCommand(c, cmPodName, cr.GetNamespace(), []string{"/bin/sh"}, command, false, false)
+	if err != nil {
+		if !mock {
+			return err
+		}
+	}
+
+	// Set cluster master maintenance mode
+	if enable {
+		cr.Status.ClusterMasterMaintenanceMode = true
+	} else {
+		cr.Status.ClusterMasterMaintenanceMode = false
+	}
+
+	return nil
+}
+
+// ApplyIdxcSecret checks if any of the indexer's have a different idxc_secret from namespace scoped secret and changes it
+func ApplyIdxcSecret(mgr *indexerClusterPodManager, n int32, mock bool) error {
```

Comment:
Nit: can we have some logical name than 'n'

Reply:
Done